In [1]:
import pandas as pd
import numpy as np
from surprise import KNNBasic, evaluate
from surprise import Dataset, Reader
from surprise import SVDpp, model_selection
from sklearn.externals import joblib
from surprise import SVD, NMF

In [2]:
us_restaurant_review_lasvegas_nv = pd.read_csv("yelp_dataset/us_restaurant_review_lasvegas_nv.csv")

In [3]:
us_restaurant_review_lasvegas_nv

Unnamed: 0             business_id  \
0              301  tstimHoMcYbkSC4eBA1wEg   
1              303  tstimHoMcYbkSC4eBA1wEg   
2              308  tstimHoMcYbkSC4eBA1wEg   
3              313  tstimHoMcYbkSC4eBA1wEg   
4              317  tstimHoMcYbkSC4eBA1wEg   
5              327  tstimHoMcYbkSC4eBA1wEg   
6              331  tstimHoMcYbkSC4eBA1wEg   
7              334  tstimHoMcYbkSC4eBA1wEg   
8              336  tstimHoMcYbkSC4eBA1wEg   
9              346  tstimHoMcYbkSC4eBA1wEg   
10             352  tstimHoMcYbkSC4eBA1wEg   
11             355  tstimHoMcYbkSC4eBA1wEg   
12             359  tstimHoMcYbkSC4eBA1wEg   
13             362  tstimHoMcYbkSC4eBA1wEg   
14             363  tstimHoMcYbkSC4eBA1wEg   
15             368  tstimHoMcYbkSC4eBA1wEg   
16             372  tstimHoMcYbkSC4eBA1wEg   
17             382  tstimHoMcYbkSC4eBA1wEg   
18             387  tstimHoMcYbkSC4eBA1wEg   
19             388  tstimHoMcYbkSC4eBA1wEg   
20             390  tstimHoMcYbkSC4eBA1wEg   
21             394  tstimHoMcYbkSC4eBA1wEg   
22             399  tstimHoMcYbkSC4eBA1wEg   
23             403  tstimHoMcYbkSC4eBA1wEg   
24             405  tstimHoMcYbkSC4eBA1wEg   
25             413  tstimHoMcYbkSC4eBA1wEg   
26             435  tstimHoMcYbkSC4eBA1wEg   
27             445  tstimHoMcYbkSC4eBA1wEg   
28             454  tstimHoMcYbkSC4eBA1wEg   
29             457  tstimHoMcYbkSC4eBA1wEg   
...            ...                     ...   
160424     3424919  ngs16C2M_uTq2zXamltHVw   
160425     3424928  ngs16C2M_uTq2zXamltHVw   
160426     3424949  ngs16C2M_uTq2zXamltHVw   
160427     3424957  ngs16C2M_uTq2zXamltHVw   
160428     3424969  ngs16C2M_uTq2zXamltHVw   
160429     3424970  ngs16C2M_uTq2zXamltHVw   
160430     3424977  ngs16C2M_uTq2zXamltHVw   
160431     3424984  ngs16C2M_uTq2zXamltHVw   
160432     3424991  ngs16C2M_uTq2zXamltHVw   
160433     3424992  ngs16C2M_uTq2zXamltHVw   
160434     3424993  ngs16C2M_uTq2zXamltHVw   
160435     3425010  ngs16C2M_uTq2zXamltHVw   
160436     3425028  ngs16C2M_uTq2zXamltHVw   
160437     3425029  ngs16C2M_uTq2zXamltHVw   
160438     3425037  ngs16C2M_uTq2zXamltHVw   
160439     3425039  ngs16C2M_uTq2zXamltHVw   
160440     3425056  ngs16C2M_uTq2zXamltHVw   
160441     3425065  ngs16C2M_uTq2zXamltHVw   
160442     3425094  ngs16C2M_uTq2zXamltHVw   
160443     3425107  ngs16C2M_uTq2zXamltHVw   
160444     3425115  ngs16C2M_uTq2zXamltHVw   
160445     3425116  ngs16C2M_uTq2zXamltHVw   
160446     3425124  ngs16C2M_uTq2zXamltHVw   
160447     3425126  ngs16C2M_uTq2zXamltHVw   
160448     3425129  ngs16C2M_uTq2zXamltHVw   
160449     3425132  ngs16C2M_uTq2zXamltHVw   
160450     3425136  ngs16C2M_uTq2zXamltHVw   
160451     3425140  ngs16C2M_uTq2zXamltHVw   
160452     3425142  ngs16C2M_uTq2zXamltHVw   
160453     3425157  ngs16C2M_uTq2zXamltHVw   

                                       name                  address  \
0       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
1       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
2       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
3       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
4       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
5       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
6       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
7       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
8       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
9       Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
10      Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
11      Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
12      Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
13      Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
14      Maria's Mexican Restaurant & Bakery    6055 E Lake Mead Blvd   
15 

# Cosine similarity

In [4]:
reader = Reader(rating_scale = (1, 5))
us_restaurant_review_lasvegas_nv_data = Dataset.load_from_df(us_restaurant_review_lasvegas_nv[['user_id', 'business_id', 'review_stars']], reader)
lasvegas_nv_data_train = us_restaurant_review_lasvegas_nv_data.build_full_trainset()
sim_options = {
    'name': 'cosine',
    'user_based': False
}

knn = KNNBasic(sim_options = sim_options)

knn.fit(lasvegas_nv_data_train)

Computing the cosine similarity matrix...
Done computing similarity matrix.


In [5]:
lasvegas_nv_data_test = lasvegas_nv_data_train.build_anti_testset()
predictions = knn.test(lasvegas_nv_data_test)

In [6]:
predictions

[Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='6fPQJq4f_yiq1NHn0fd11Q', r_ui=3.774103481371608, est=3.875, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='k-dDZvTeLysoJvjHI-qr9g', r_ui=3.774103481371608, est=3.775, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='eY98EH_PHGnJVaJ7t2UpoQ', r_ui=3.774103481371608, est=4.175, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='wdsGP5rJinx0J5GcGzPQUw', r_ui=3.774103481371608, est=3.875, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='k2b3niokS_tosjah_rzCPw', r_ui=3.774103481371608, est=3.7, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='dusNIzdCaH6EoLl2hRy6cQ', r_ui=3.774103481371608, est=4.025, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='h

# Matrix Factorization

In [7]:
svdpp = SVDpp()
svdpp.fit(lasvegas_nv_data_train)

In [8]:
predictions_svdpp = svdpp.predict('3nDUQBjKyVor5wV0reJChg', '6fPQJq4f_yiq1NHn0fd11Q')
predictions_svdpp

Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='6fPQJq4f_yiq1NHn0fd11Q', r_ui=None, est=4.036523602416956, details={'was_impossible': False})

### Grab business_ids that user_id 3nDUQBjKyVor5wV0reJChg did not rate in order to recommend restaurants they have not visited


In [9]:
business_ids = us_restaurant_review_lasvegas_nv['business_id'].unique()
business_ids_rated_by_user = us_restaurant_review_lasvegas_nv.loc[us_restaurant_review_lasvegas_nv['user_id'] == '3nDUQBjKyVor5wV0reJChg', 'business_id']
business_ids_to_predict = np.setdiff1d(business_ids, business_ids_rated_by_user)

### Predict the review_ratings for each business_id that user_id 3nDUQBjKyVor5wV0reJChg did not rate

In [10]:
testset = [['3nDUQBjKyVor5wV0reJChg', business_id, 4.] for business_id in business_ids_to_predict]
predictions_testset = svdpp.test(testset)
predictions_testset

[Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='--9e1ONYQuAa-CB_Rrw7Tw', r_ui=4.0, est=4.010950843454543, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-0RkJ_uIduNLWQrphbADRw', r_ui=4.0, est=3.3695078444349194, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-1m9o3vGRA8IBPNvNqKLmA', r_ui=4.0, est=4.116472496835979, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-1vfRrlnNnNJ5boOVghMPA', r_ui=4.0, est=2.2416544940314225, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-3zffZUHoY8bQjGfPSoBKQ', r_ui=4.0, est=3.603354445929827, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-8R_-EkGpUhBk55K9Dd4mg', r_ui=4.0, est=2.9563177580454583, details={'was_impossible': False}),
 Prediction(uid='3nDUQBjKyVor5wV0reJChg', iid='-9YyInW1wapzdNZrhQJ9dg', r_ui=4.0, est=2.051952463825889, details={'was_impossible': False}),
 Predictio

### Output the top 30 business_ids with the top 30 best predicted review_rating

In [11]:
predicted_ratings = np.array([pred.est for pred in predictions_testset])
index_max_pred_rating = np.argpartition(predicted_ratings,-30)[-30:]
business_id_recommended = business_ids_to_predict[index_max_pred_rating]
print('Top 10 businesses:\n {0}, \npredicted review_ratings: \n{1}'.format(business_id_recommended, predicted_ratings[index_max_pred_rating]))

Top 10 businesses:
 ['Rd11Bosr8JkfoUhZLzpuxg' '7hWNnAj4VwK6FAUBN8E8lg'
 'RwMLuOkImBIqqYj4SSKSPg' 'H8qpFitIesth86zqR4cwYg'
 'O19VReN1I2TBrJsbXUAIJg' 'vLkKorA1gPS97Zj0_WKA0Q'
 'd_L-rfS1vT3JMzgCUGtiow' 'kafUXRNfQm3e_U14S2WbmA'
 '8SwMaPL6AJIs1ZPSZ3NJ4w' '2B46bRpDh49eDyjXGhL_ZQ'
 'D6K3VT6S8FwFm4u5uN8T8g' '3XE-poXz9CeDao-Q7pqkTw'
 'cfggUk66F01Av4YWZJXcnA' 'pyqnGlIfP9Zw8LLxBjyAiA'
 'G4hjhtA_wQ-tSOGpgGlDjw' 'vT812nRctS62PVxUverayw'
 '3pSUr_cdrphurO6m1HMP9A' 'LRpFWbovjUShlvCKQOS-Ng'
 '6qKjl6xJQzLb9ka-QbTudA' '2hSIeCX2cV-chFcBBXrZZA'
 'G-5kEa6E6PD5fkBRuA7k9Q' 'RNi6tW22UMgHwWLAb0mYdA'
 'lQ9VmIWvM6v-rONPkYX8aQ' 'J4CATH00YZrq8Bne2S4_cw'
 'S-egzF2c7bRpCUNbZqV8dg' 'TCWMgOiV0PxQkWE1SyBIWQ'
 'w8RWhhuKxMiFCpvFvIQ3zg' 'egpDg61SdYbKWvCpPM7Nbg'
 'OETh78qcgDltvHULowwhJg' 'mUdalYTuAtnZm2K9zoH27Q'], 
predicted review_ratings: 
[4.59260675 4.60402946 4.6186503  4.60430196 4.62041257 4.62140496
 4.66357655 4.69561608 4.67211946 4.75162772 4.70767504 4.69532577
 4.6974384  4.64413013 4.69973416 4.77748642 4.8041

# Untuned Models

## Mean RMSE values
### SVDpp: 1.0095
### SVD: 1.0065
### NMF: 1.0545

## Since SVDpp and SVD have lower RMSE values than NMF, we would tune only SVDpp and SVD as they have greater potential to produce even lower RMSE values, indicating a more accurate model and hence predictions.

In [23]:
svdpp = SVDpp()
model_selection.cross_validate(svdpp, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0138  1.0067  1.0060  1.0072  1.0135  1.0095  0.0035  
MAE (testset)     0.7884  0.7831  0.7830  0.7819  0.7902  0.7853  0.0033  
Fit time          138.41  140.14  141.62  144.71  144.57  141.89  2.46    
Test time         2.77    2.79    2.80    3.03    2.95    2.87    0.10    


{'test_rmse': array([1.0138335 , 1.00666495, 1.00598413, 1.00723393, 1.01353477]),
 'test_mae': array([0.7883986 , 0.78313281, 0.78303952, 0.78192596, 0.79017396]),
 'fit_time': (138.40799379348755,
  140.13824725151062,
  141.61850905418396,
  144.70594954490662,
  144.56979656219482),
 'test_time': (2.7745842933654785,
  2.7895450592041016,
  2.8045854568481445,
  3.0328941345214844,
  2.947873115539551)}

In [24]:
svd = SVD()
model_selection.cross_validate(svd, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0042  1.0056  0.9986  1.0123  1.0119  1.0065  0.0051  
MAE (testset)     0.7823  0.7844  0.7770  0.7891  0.7890  0.7843  0.0045  
Fit time          6.35    6.07    6.14    6.17    6.31    6.21    0.11    
Test time         0.21    0.17    0.17    0.17    0.18    0.18    0.01    


{'test_rmse': array([1.00417675, 1.00555005, 0.99861719, 1.0123166 , 1.01188611]),
 'test_mae': array([0.78226978, 0.78435236, 0.77699511, 0.78905269, 0.78899995]),
 'fit_time': (6.353667974472046,
  6.067141532897949,
  6.14189887046814,
  6.166175842285156,
  6.306578874588013),
 'test_time': (0.20641756057739258,
  0.17453432083129883,
  0.17153406143188477,
  0.17453384399414062,
  0.17952823638916016)}

In [25]:
nmf = NMF()
model_selection.cross_validate(nmf, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0465  1.0625  1.0537  1.0547  1.0549  1.0545  0.0051  
MAE (testset)     0.8135  0.8241  0.8198  0.8183  0.8172  0.8186  0.0035  
Fit time          7.00    6.95    6.69    7.02    6.68    6.87    0.15    
Test time         0.15    0.16    0.14    0.14    0.14    0.15    0.01    


{'test_rmse': array([1.04647433, 1.06253072, 1.05368172, 1.05472114, 1.05490566]),
 'test_mae': array([0.81349687, 0.8241269 , 0.819798  , 0.81832553, 0.81722433]),
 'fit_time': (7.002248525619507,
  6.952529430389404,
  6.691587448120117,
  7.015589714050293,
  6.68014931678772),
 'test_time': (0.14983773231506348,
  0.15558552742004395,
  0.13962697982788086,
  0.13866233825683594,
  0.14458012580871582)}

# Model Tuning and Evaluation

### SVDpp

#### 'lr_all':[.001,.01], 'reg_all':[.1,.5] cv=3

In [49]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.1}


In [50]:
svdpp_tuned = SVDpp(lr_all= .01, reg_all = .1)
model_selection.cross_validate(svdpp_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0008  0.9929  0.9983  1.0030  0.9974  0.9985  0.0034  
MAE (testset)     0.7811  0.7763  0.7785  0.7827  0.7772  0.7792  0.0024  
Fit time          143.65  149.46  147.60  151.56  148.55  148.16  2.61    
Test time         2.66    2.74    2.70    2.62    2.60    2.66    0.05    


{'test_rmse': array([1.00083095, 0.99293023, 0.9983339 , 1.00295769, 0.99743116]),
 'test_mae': array([0.78107717, 0.77628341, 0.7784986 , 0.78270869, 0.77718406]),
 'fit_time': (143.6495156288147,
  149.4600739479065,
  147.59630274772644,
  151.56176257133484,
  148.54871559143066),
 'test_time': (2.6568753719329834,
  2.7379534244537354,
  2.696866273880005,
  2.6170029640197754,
  2.5960583686828613)}

### When cv is increased to 5, the lr_all and reg_all with the best RMSE score is still the same.
#### 'lr_all':[.001,.01], 'reg_all':[.1,.5] cv=5

In [12]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=5)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.1}


### When reg_all range for testing is increased from .1-.5 to .01-.5, the reg_all that produced the best RMSE score changes to 0.5. 
### However, building the model with those parameters shows that the mean RMSE is higher at 1.0070 when reg_all = .5, compared to the mean RMSE at 0.9985 when reg_all = .1
#### 'lr_all':[.001,.01], 'reg_all':[.01,.5] cv=5

In [13]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.01,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=5)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.5}


In [14]:
svdpp_tuned_2 = SVDpp(lr_all= .01, reg_all = .5)
model_selection.cross_validate(svdpp_tuned_2, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0024  1.0055  1.0168  1.0012  1.0092  1.0070  0.0056  
MAE (testset)     0.7901  0.7911  0.7997  0.7879  0.7926  0.7923  0.0040  
Fit time          143.23  140.90  143.84  142.15  150.13  144.05  3.20    
Test time         2.85    2.76    3.01    3.32    3.12    3.01    0.20    


{'test_rmse': array([1.0024019 , 1.00545414, 1.01675473, 1.00124677, 1.00917391]),
 'test_mae': array([0.79009848, 0.79112877, 0.79965108, 0.78794482, 0.79260655]),
 'fit_time': (143.23379635810852,
  140.89931440353394,
  143.83815503120422,
  142.1469748020172,
  150.12847352027893),
 'test_time': (2.8487355709075928,
  2.764979362487793,
  3.010981798171997,
  3.318101644515991,
  3.121657371520996)}

### No changes observed to the best parameters when reg_all range is changed from .1-.5 to .01-.1

In [15]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.01,.1]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=5)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.1}


### Lower mean RMSE value of 0.9979 when lr_all = 0.005 & reg_all = 0.1, compared to mean RMSE value of 0.9985 when lr_all = 0.01 & reg_all = 0.1

In [19]:
parameters_grid = {'lr_all':[.001,.005], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.005, 'reg_all': 0.1}


In [20]:
svdpp_tuned = SVDpp(lr_all= .005, reg_all = .1)
model_selection.cross_validate(svdpp_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9979  0.9980  1.0007  0.9971  0.9958  0.9979  0.0016  
MAE (testset)     0.7813  0.7790  0.7815  0.7806  0.7772  0.7799  0.0016  
Fit time          150.48  144.86  141.41  145.37  140.59  144.54  3.51    
Test time         2.86    2.79    3.04    2.79    3.05    2.91    0.12    


{'test_rmse': array([0.99794678, 0.99796958, 1.0006837 , 0.99707348, 0.99584939]),
 'test_mae': array([0.78125303, 0.77904245, 0.78149572, 0.7805548 , 0.77721775]),
 'fit_time': (150.47993302345276,
  144.85952305793762,
  141.4088213443756,
  145.36533188819885,
  140.59009313583374),
 'test_time': (2.860353946685791,
  2.7944977283477783,
  3.0356040000915527,
  2.786555290222168,
  3.051877021789551)}

# Best performing 
### SVDpp model with lowest RMSE value of 0.9977: n_factors = 30, lr_all = 0.005, reg_all = 0.1

In [30]:
parameters_grid = {'lr_all':[.001,.005], 'reg_all':[.1,.5], 'n_factors':[10,30] }
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all, reg_all and n_factors that produced best RMSE score

{'lr_all': 0.005, 'reg_all': 0.1, 'n_factors': 30}


In [31]:
svdpp_tuned = SVDpp(lr_all= .005, reg_all = .1, n_factors=30)
model_selection.cross_validate(svdpp_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9905  0.9924  1.0051  1.0003  0.9999  0.9977  0.0054  
MAE (testset)     0.7743  0.7753  0.7844  0.7820  0.7833  0.7799  0.0042  
Fit time          175.90  175.66  176.08  176.89  178.57  176.62  1.06    
Test time         2.84    2.83    2.80    2.87    2.86    2.84    0.03    


{'test_rmse': array([0.99053121, 0.99241859, 1.00514135, 1.00033849, 0.99992502]),
 'test_mae': array([0.77429846, 0.77529318, 0.78440009, 0.7819775 , 0.78330846]),
 'fit_time': (175.90186309814453,
  175.66474199295044,
  176.07554364204407,
  176.885178565979,
  178.56954669952393),
 'test_time': (2.843071222305298,
  2.8274497985839844,
  2.796170711517334,
  2.874328851699829,
  2.8599319458007812)}

### Higher mean RMSE value of 0.9987 when lr_all = 0.004 & reg_all = 0.1, compared to mean RMSE value of 0.9985 when lr_all = 0.01 & reg_all = 0.1

In [21]:
parameters_grid = {'lr_all':[.001,.004], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.004, 'reg_all': 0.1}


In [22]:
svdpp_tuned = SVDpp(lr_all= .004, reg_all = .1)
model_selection.cross_validate(svdpp_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0039  0.9913  0.9982  1.0005  0.9994  0.9987  0.0041  
MAE (testset)     0.7842  0.7770  0.7807  0.7822  0.7812  0.7811  0.0024  
Fit time          221.91  139.46  142.07  139.68  142.17  157.06  32.44   
Test time         2.97    2.77    2.82    2.79    2.92    2.85    0.08    


{'test_rmse': array([1.00388987, 0.99129037, 0.99821793, 1.0005187 , 0.99944177]),
 'test_mae': array([0.78423999, 0.77699982, 0.7807059 , 0.78218117, 0.78118986]),
 'fit_time': (221.9050109386444,
  139.45560932159424,
  142.07469272613525,
  139.67606616020203,
  142.17255568504333),
 'test_time': (2.9683146476745605,
  2.7676055431365967,
  2.819464683532715,
  2.7937886714935303,
  2.9231889247894287)}

### SVD

In [16]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.01,.5]}
gridsearch = model_selection.GridSearchCV(SVD, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.5}


In [17]:
parameters_grid = {'lr_all':[.001,.01], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVDpp, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.01, 'reg_all': 0.1}


In [18]:
svd_tuned = SVD(lr_all= .01, reg_all = .1)
model_selection.cross_validate(svd_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9980  0.9934  1.0024  0.9978  1.0051  0.9993  0.0041  
MAE (testset)     0.7776  0.7759  0.7831  0.7778  0.7820  0.7793  0.0028  
Fit time          5.99    5.99    5.93    6.02    6.00    5.99    0.03    
Test time         0.17    0.17    0.17    0.16    0.17    0.17    0.01    


{'test_rmse': array([0.99801253, 0.99340397, 1.00242163, 0.9977638 , 1.00509735]),
 'test_mae': array([0.77760165, 0.77586218, 0.78309493, 0.7777755 , 0.78196985]),
 'fit_time': (5.994533538818359,
  5.988775730133057,
  5.9339423179626465,
  6.016851186752319,
  5.9974541664123535),
 'test_time': (0.17183184623718262,
  0.17183494567871094,
  0.17186808586120605,
  0.15528512001037598,
  0.16953086853027344)}

In [26]:
parameters_grid = {'lr_all':[.001,.005], 'reg_all':[.1,.5]}
gridsearch = model_selection.GridSearchCV(SVD, parameters_grid, measures=['rmse','mae'], cv=3)
gridsearch.fit(us_restaurant_review_lasvegas_nv_data)
print(gridsearch.best_params['rmse']) #combination of lr_all and reg_all that produced best RMSE score

{'lr_all': 0.005, 'reg_all': 0.1}


In [27]:
svd_tuned = SVD(lr_all= .005, reg_all = .1)
model_selection.cross_validate(svd_tuned, us_restaurant_review_lasvegas_nv_data, verbose = True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0006  0.9990  0.9982  0.9983  0.9939  0.9980  0.0022  
MAE (testset)     0.7813  0.7808  0.7800  0.7818  0.7764  0.7800  0.0019  
Fit time          6.10    6.30    6.04    6.07    6.17    6.14    0.09    
Test time         0.18    0.17    0.17    0.18    0.21    0.18    0.01    


{'test_rmse': array([1.00062931, 0.9989689 , 0.99816299, 0.99825001, 0.99390977]),
 'test_mae': array([0.78125654, 0.78084539, 0.78000039, 0.78177964, 0.77636553]),
 'fit_time': (6.0960776805877686,
  6.3011109828948975,
  6.042832374572754,
  6.073749780654907,
  6.1655592918396),
 'test_time': (0.18262219429016113,
  0.17257213592529297,
  0.17453479766845703,
  0.179487943649292,
  0.21143507957458496)}

### Serialize Models

In [19]:
joblib.dump(knn, 'las_vegas_existing_user_model.pkl')

['las_vegas_existing_user_model.pkl']

In [53]:
joblib.dump(svdpp, 'svdpp_las_vegas_existing_user_model.pkl')

['svdpp_las_vegas_existing_user_model.pkl']

In [8]:
# Split us_restaurant_review_lasvegas_nv_data into 5 folds
from surprise import SVD, evaluate
from surprise import NMF

us_restaurant_review_lasvegas_nv_data.split(n_folds=5)

svd = SVD()
evaluate(svd, us_restaurant_review_lasvegas_nv_data, measures=['RMSE'])

svdpp = SVDpp()
evaluate(svdpp, us_restaurant_review_lasvegas_nv_data, measures=['RMSE'])

nmf = NMF()
evaluate(nmf, us_restaurant_review_lasvegas_nv_data, measures=['RMSE'])

Evaluating RMSE of algorithm SVD.

------------
Fold 1
RMSE: 1.0016
------------
Fold 2
RMSE: 1.0091
------------
Fold 3
RMSE: 1.0051
------------
Fold 4
RMSE: 1.0104
------------
Fold 5
RMSE: 1.0074
------------
------------
Mean RMSE: 1.0067
------------
------------
Evaluating RMSE of algorithm SVDpp.

------------
Fold 1
RMSE: 1.0017
------------
Fold 2
RMSE: 1.0113
------------
Fold 3
RMSE: 1.0095
------------
Fold 4
RMSE: 1.0126
------------
Fold 5
RMSE: 1.0098
------------
------------
Mean RMSE: 1.0090
------------
------------
Evaluating RMSE of algorithm NMF.

------------
Fold 1
RMSE: 1.0472
------------
Fold 2
RMSE: 1.0561
------------
Fold 3
RMSE: 1.0563
------------
Fold 4
RMSE: 1.0621
------------
Fold 5
RMSE: 1.0540
------------
------------
Mean RMSE: 1.0551
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [1.0471854044890245,
                             1.0560891605198859,
                             1.056308211405088,
                             1.0620656574715148,
                             1.0539947150759905]})

In [2]:
us_restaurant_review = pd.read_csv("yelp_dataset/us_restaurant_review.csv")

us_restaurant_review

C:\Users\yunin\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0             business_id                         name  \
0                 0  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
1                 1  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
2                 2  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
3                 3  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
4                 4  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
5                 5  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
6                 6  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
7                 7  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
8                 8  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
9                 9  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
10               10  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
11               11  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
12               12  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
13               13  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
14               14  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
15               15  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
16               16  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
17               17  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
18               18  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
19               19  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
20               20  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
21               21  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
22               22  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
23               23  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
24               24  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
25               25  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
26               26  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
27               27  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
28               28  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
29               29  gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
...             ...                     ...                          ...   
3425371     3425371  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425372     3425372  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425373     3425373  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425374     3425374  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425375     3425375  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425376     3425376  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425377     3425377  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425378     3425378  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425379     3425379  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425380     3425380  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425381     3425381  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425382     3425382  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425383     3425383  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425384     3425384  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425385     3425385  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425386     3425386  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425387     3425387  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425388     3425388  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425389     3425389  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425390     3425390  p1eUTUB_5hf1enoKYBV3DA      Ken Stewart's Tre Belle   
3425391     3425391  p1eUTUB_5hf1enoKYBV3DA      Ken Stew

In [3]:
us_restaurant_review.dtypes

Unnamed: 0            int64
business_id          object
name                 object
address              object
city                 object
state                object
postal_code          object
latitude            float64
longitude           float64
restaurant_stars    float64
review_count          int64
categories           object
review_id            object
user_id              object
review_stars        float64
date                 object
text                 object
useful                int64
funny                 int64
cool                  int64
dtype: object

In [4]:
us_restaurant_review.shape

(3425401, 20)

### Even after reducing the number of rows in dataset to 100000, we still ran into memory error.
### This could have been due to the sparsity of the matrix.

In [5]:
us_test = us_restaurant_review.head(100000)
us_test.shape

(100000, 20)

In [6]:
reader = Reader(rating_scale = (1, 5))
us_test_data = Dataset.load_from_df(us_test[['user_id', 'business_id', 'review_stars']], reader)

In [7]:
us_test_data

In [8]:
data_train = us_test_data.build_full_trainset()

In [9]:
sim_options = {
    'name': 'cosine',
    'user_based': False
}

knn = KNNBasic(sim_options = sim_options)

knn.fit(data_train)

Computing the cosine similarity matrix...
Done computing similarity matrix.


In [37]:
data_test = data_train.build_anti_testset()
predictions = knn.test(data_test)

MemoryError: 

### Successful generation of predictions when number of rows in dataset is reduced to 10000.
### However, the model produced impossible predictions, as they were random. Possible reasons for impossible predictions: Either the size of the dataset was not large enough for the predictions to be statistically significant, or the matrix was too sparse.

In [10]:
us_test = us_restaurant_review.head(10000)
reader = Reader(rating_scale = (1, 5))
us_test_data = Dataset.load_from_df(us_test[['user_id', 'business_id', 'review_stars']], reader)
data_train = us_test_data.build_full_trainset()
sim_options = {
    'name': 'cosine',
    'user_based': False
}

knn = KNNBasic(sim_options = sim_options)

knn.fit(data_train)

Computing the cosine similarity matrix...
Done computing similarity matrix.


In [11]:
data_test = data_train.build_anti_testset()
predictions = knn.test(data_test)

In [12]:
predictions

[Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='1Dfx3zM-rW4n-31KeC8sJg', r_ui=3.7122, est=3.7122, details={'was_impossible': True, 'reason': 'Not enough neighbors.'}),
 Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='fweCYi8FmbJXHCqLnwuk8w', r_ui=3.7122, est=3.7122, details={'was_impossible': True, 'reason': 'Not enough neighbors.'}),
 Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='PZ-LZzSlhSe9utkQYU8pFg', r_ui=3.7122, est=3.7122, details={'was_impossible': True, 'reason': 'Not enough neighbors.'}),
 Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='1RHY4K3BD22FK7Cfftn8Mg', r_ui=3.7122, est=3.7122, details={'was_impossible': True, 'reason': 'Not enough neighbors.'}),
 Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='tstimHoMcYbkSC4eBA1wEg', r_ui=3.7122, est=3.7122, details={'was_impossible': True, 'reason': 'Not enough neighbors.'}),
 Prediction(uid='nsS4oDfOsl20QdWc6XcOkw', iid='BvYU3jvGd0TJ7IyZdfiN2Q', r_ui=3.7122, est=2.0, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='nsS